MergeContext.js

```diff
@@ -0,0 +1,506 @@
+    // Label manipulation methods
+
+    async _hasLabel(label) {
+        const labels = await GH.getLabels(this._number());
+        return labels.find(lbl => lbl.name === label) !== undefined;
+    }
+
+    async _removeLabel(label) {
+        try {
+            await GH.removeLabel(label, this._number());
+        } catch (e) {
+            if (e.name === 'ErrorContext' && e.notFound()) {
+                this._log("removeLabel: " + label + " not found");
+                return;
+            }
+            throw e;
+        }
+    }
+
+    async _addLabel(label) {
+        let params = Util.commonParams();
+        params.number = this._number();
+        params.labels = [];
+        params.labels.push(label);
+        try {
+            await GH.addLabels(params);
+        } catch (e) {
+            // TODO: also extract and check for "already_exists" code:
+            // { "message": "Validation Failed", "errors": [ { "resource": "Label", "code": "already_exists", "field": "name" } ] }
+            if (e.name === 'ErrorContext' && e.unprocessable()) {
+                this._log("addLabel: " + label + " already exists");
+                return;
+            }
+            throw e;
+        }
+    }
+
+    async _labelFailedDescription(isValid) {
+        const label = Config.failedDescriptionLabel();
+        if (isValid)
+            await this._removeLabel(label);
+        else
+            await this._addLabel(label);
+    }
+
+    async _labelWaitingStagingChecks() {
+        await this._removeLabel(Config.passedStagingChecksLabel());
+        await this._removeLabel(Config.failedOtherLabel());
+        await this._removeLabel(Config.failedStagingChecksLabel());
+        await this._addLabel(Config.waitingStagingChecksLabel());
+    }
+
+    async _labelMerged() {
+        await this._removeLabel(Config.waitingStagingChecksLabel());
+        await this._removeLabel(Config.passedStagingChecksLabel());
+        await this._addLabel(Config.mergedLabel());
+    }
+
+    async _labelFailedOther() {
+        await this._removeLabel(Config.waitingStagingChecksLabel());
+        await this._removeLabel(Config.passedStagingChecksLabel());
+        await this._addLabel(Config.failedOtherLabel());
+    }
+
+    async _labelFailedStagingChecks() {
+        await this._removeLabel(Config.waitingStagingChecksLabel());
+        await this._addLabel(Config.failedStagingChecksLabel());
+    }
+
+    async _labelPassedStagingChecks() {
+        await this._removeLabel(Config.waitingStagingChecksLabel());
+        await this._addLabel(Config.passedStagingChecksLabel());
+    }
+
+    // Getters
+
+    // the processing of this PR is delayed on this
+    // number of milliseconds
+    delay() { return this._votingDelay; }
+
+    _number() { return this._pr.number; }
+
+    _prHeadSha() { return this._pr.head.sha; }
+
+    _prMessage() {
+        return this._pr.title + endOfLine + this._pr.body + endOfLine +
+            "(PR #" + this._pr.number + ")";
```

Comment:
Drop the `PR` part. Again, see actual GitHub commits for examples.

Reply:
Done.